web/src/typeahead_helper.ts

```diff
@@ -502,16 +502,40 @@ export let sort_recipients = <UserType extends UserOrMentionPillData | UserPillD
     ];
     const groups_okay_matches = [...groups_results.word_boundary_matches];
 
-    const best_users = (): UserType[] => [
+    const best_users_including_bots = (): UserType[] => [
         ...sort_relevance(users_name_good_matches),
         ...sort_relevance(users_name_okay_matches),
     ];
+
+    function filter_bot_user(user: UserType, is_bot: boolean): boolean {
+        if (user.type === "broadcast") {
+            // broadcast items are not bots by definition.
+            return !is_bot;
+        }
+        return user.user.is_bot === is_bot;
+    }
+
+    const best_users = (): UserType[] =>
+        best_users_including_bots().filter((user) => filter_bot_user(user, false));
+    const best_bots = (): UserType[] =>
+        best_users_including_bots().filter((user) => filter_bot_user(user, true));
```

Comment:
I think there's probably a better JS API for splitting a list into two by a filtering function that doesn't involve iterating through the whole list twice. Not sure it matters much. This is a highly performance-sensitive code path, but all of `sort_recipients` is small compared to template rendering right now.

Reply:
Okay, used a different strategy.